File: server/routers/rag_chat.py
"""
RAG-grounded chat endpoint for the Zena widget.

Endpoint:
  POST /v1/rag/chat  — query → RAG search → context inject → LLM generate → respond

This is the backend that powers the embedded Zena chatbot widget.
It retrieves relevant context from the RAG pipeline before generating a response.
"""

import asyncio
import logging
import time
import uuid
from typing import Any, Dict, List, Optional

from fastapi import APIRouter, HTTPException
from fastapi.responses import JSONResponse
from pydantic import BaseModel

logger = logging.getLogger("api_server")

router = APIRouter()


# ── Request / Response schemas ────────────────────────────────────────


class RAGChatMessage(BaseModel):
    role: str = "user"
    content: str = ""


class RAGChatRequest(BaseModel):
    """Request body for /v1/rag/chat."""

    messages: List[RAGChatMessage] = []
    temperature: float = 0.7
    max_tokens: Optional[int] = 512
    top_k: int = 5  # RAG retrieval depth
    score_threshold: float = 0.3  # minimum relevance
    system_prompt: Optional[str] = None  # override system prompt
    focus_mode: Optional[str] = None  # prompt focus mode (e.g. "data_extraction")


class RAGChatResponse(BaseModel):
    id: str
    object: str = "rag.chat.completion"
    created: int
    model: str = "rag-pipeline"
    choices: List[Dict[str, Any]]
    usage: Dict[str, Any] = {}
    rag_sources: List[Dict[str, Any]] = []  # sources used for context


# ── Singleton accessor for the RAG integration ────────────────────────


def _get_rag_integration():
    """Try to import and return the global RAG integration instance."""
    try:
        from rag_integration import RAGIntegration

        return RAGIntegration()
    except Exception:
        return None


def _get_llm_state():
    """Get the API server state (LLM adapter) if available."""
    try:
        from server.helpers import get_state

        return get_state()
    except Exception:
        return None


# ── The endpoint ──────────────────────────────────────────────────────


@router.post("/v1/rag/chat")
async def rag_chat(req: RAGChatRequest):
    """
    RAG-augmented chat completion.

    1. Extract the user query from the last message
    2. Search the RAG knowledge base for relevant context
    3. Inject context into the system prompt
    4. Generate a response via the LLM (or return context-only if no LLM)
    """
    if not req.messages:
        raise HTTPException(400, detail="messages required")

    # Extract the latest user message as the query
    user_query = ""
    for msg in reversed(req.messages):
        if msg.role == "user" and msg.content.strip():
            user_query = msg.content.strip()
            break
    if not user_query:
        raise HTTPException(400, detail="No user message found")

    request_id = f"ragchat-{uuid.uuid4().hex[:12]}"
    created = int(time.time())

    # ── Step 1: RAG retrieval ─────────────────────────────────────────
    rag = _get_rag_integration()
    context_text = ""
    rag_sources: List[Dict[str, Any]] = []

    if rag and rag.initialized:
        try:
            context_text, raw_results = await rag.query_context(
                query=user_query,
                top_k=req.top_k,
                score_threshold=req.score_threshold,
            )
            rag_sources = [
                {
                    "text": r.get("text", "")[:300],
                    "score": round(r.get("score", 0), 3),
                    "source": r.get("source", "unknown"),
                }
                for r in raw_results
            ]
        except Exception as e:
            logger.warning(f"[rag-chat] RAG retrieval failed: {e}")
            context_text = ""
            rag_sources = []

    # ── Step 2: Build messages with context ──────────────────────────
    # Apply focus mode if specified (prompt injection for data tasks)
    focused_query = user_query
    try:
        from Core.prompt_focus import FocusMode, apply_focus

        if req.focus_mode:
            _mode = FocusMode.from_string(req.focus_mode)
            if _mode != FocusMode.GENERAL:
                _focus_sys, focused_query = apply_focus(_mode, user_query, req.system_prompt)
                if not req.system_prompt:
                    req.system_prompt = _focus_sys
    except ImportError as exc:
        logger.debug("%s", exc)

    system_prompt = req.system_prompt or (
        "You are Zena, a helpful virtual assistant. "
        "Answer the user's question using ONLY the provided context. "
        "If the context doesn't contain the answer, say so honestly. "
        "Respond in the same language the user uses."
    )

    if context_text:
        system_prompt += f"\n\n--- KNOWLEDGE BASE CONTEXT ---\n{context_text}\n--- END CONTEXT ---"

    messages_for_llm = [{"role": "system", "content": system_prompt}]
    for msg in req.messages:
        # Replace the last user message with the focused version
        if msg.role == "user" and msg.content.strip() == user_query:
            messages_for_llm.append({"role": msg.role, "content": focused_query})
        else:
            messages_for_llm.append({"role": msg.role, "content": msg.content})

    # ── Step 3: Try LLM generation ───────────────────────────────────
    state = _get_llm_state()
    response_text = ""

    if state and state.ready and state.adapter:
        try:
            from server.helpers import build_inference_request, estimate_tokens
            from server.schemas import ChatCompletionRequest, ChatMessage

            # Build a ChatCompletionRequest for the existing infrastructure
            chat_req = ChatCompletionRequest(
                messages=[ChatMessage(role=m["role"], content=m["content"]) for m in messages_for_llm],
                temperature=req.temperature,
                max_tokens=req.max_tokens or 512,
                stream=False,
            )
            llm_req = build_inference_request(messages_for_llm, chat_req)

            async with state.inference_semaphore:
                result = await asyncio.to_thread(state.adapter.generate, llm_req)
            response_text = result if isinstance(result, str) else str(result)

            # Cache the result
            try:
                state.cache.put(
                    messages_for_llm,
                    req.temperature,
                    req.max_tokens or 512,
                    response_text,
                )
            except Exception as exc:
                logger.debug("%s", exc)

            prompt_tokens = estimate_tokens(messages_for_llm)
            completion_tokens = estimate_tokens(response_text)
            state.record_request(completion_tokens)

            return JSONResponse(
                content={
                    "id": request_id,
                    "object": "rag.chat.completion",
                    "created": created,
                    "model": getattr(state, "model_id", "rag-pipeline"),
                    "choices": [
                        {
                            "index": 0,
                            "message": {
                                "role": "assistant",
                                "content": response_text,
                            },
                            "finish_reason": "stop",
                        }
                    ],
                    "usage": {
                        "prompt_tokens": prompt_tokens,
                        "completion_tokens": completion_tokens,
                        "total_tokens": prompt_tokens + completion_tokens,
                    },
                    "rag_sources": rag_sources,
                }
            )

        except Exception as e:
            logger.warning(f"[rag-chat] LLM generation failed: {e}, falling back to context-only")

    # ── Fallback: return RAG context without LLM ─────────────────────
    if context_text:
        # Format context as a readable answer
        response_text = "Based on the knowledge base, here is what I found:\n\n" + context_text[:2000]
    else:
        response_text = (
            "I don't have enough information to answer that question yet. "
            "Please make sure data has been loaded into the RAG pipeline."
        )

    return JSONResponse(
        content={
            "id": request_id,
            "object": "rag.chat.completion",
            "created": created,
            "model": "rag-context-only",
            "choices": [
                {
                    "index": 0,
                    "message": {
                        "role": "assistant",
                        "content": response_text,
                    },
                    "finish_reason": "stop",
                }
            ],
            "usage": {},
            "rag_sources": rag_sources,
        }
    )

File: rust_output/src/rag_chat.rs
/// RAG-grounded chat endpoint for the Zena widget.
/// 
/// Endpoint:
/// POST /v1/rag/chat  — query → RAG search → context inject → LLM generate → respond
/// 
/// This is the backend that powers the embedded Zena chatbot widget.
/// It retrieves relevant context from the RAG pipeline before generating a response.

use anyhow::{Result, Context};
use std::collections::HashMap;
use tokio;

pub static LOGGER: std::sync::LazyLock<String /* logging::getLogger */> = std::sync::LazyLock::new(|| Default::default());

pub static ROUTER: std::sync::LazyLock<APIRouter> = std::sync::LazyLock::new(|| Default::default());

#[derive(Debug, Clone)]
pub struct RAGChatMessage {
    pub role: String,
    pub content: String,
}

/// Request body for /v1/rag/chat.
#[derive(Debug, Clone)]
pub struct RAGChatRequest {
    pub messages: Vec<RAGChatMessage>,
    pub temperature: f64,
    pub max_tokens: Option<i64>,
    pub top_k: i64,
    pub score_threshold: f64,
    pub system_prompt: Option<String>,
    pub focus_mode: Option<String>,
}

#[derive(Debug, Clone)]
pub struct RAGChatResponse {
    pub id: String,
    pub object: String,
    pub created: i64,
    pub model: String,
    pub choices: Vec<HashMap<String, Box<dyn std::any::Any>>>,
    pub usage: HashMap<String, Box<dyn std::any::Any>>,
    pub rag_sources: Vec<HashMap<String, Box<dyn std::any::Any>>>,
}

/// Try to import and return the global RAG integration instance.
pub fn _get_rag_integration() -> Result<()> {
    // Try to import and return the global RAG integration instance.
    // try:
    {
        // TODO: from rag_integration import RAGIntegration
        RAGIntegration()
    }
    // except Exception as _e:
}

/// Get the API server state (LLM adapter) if available.
pub fn _get_llm_state() -> Result<()> {
    // Get the API server state (LLM adapter) if available.
    // try:
    {
        // TODO: from server::helpers import get_state
        get_state()
    }
    // except Exception as _e:
}

/// RAG-augmented chat completion.
/// 
/// 1. Extract the user query from the last message
/// 2. Search the RAG knowledge base for relevant context
/// 3. Inject context into the system prompt
/// 4. Generate a response via the LLM (or return context-only if no LLM)
pub async fn rag_chat(req: RAGChatRequest) -> Result<()> {
    // RAG-augmented chat completion.
    // 
    // 1. Extract the user query from the last message
    // 2. Search the RAG knowledge base for relevant context
    // 3. Inject context into the system prompt
    // 4. Generate a response via the LLM (or return context-only if no LLM)
    if !req.messages {
        return Err(anyhow::anyhow!("HTTPException(400, detail='messages required')"));
    }
    let mut user_query = "".to_string();
    for msg in req.messages.iter().rev().iter() {
        if (msg.role == "user".to_string() && msg.content.trim().to_string()) {
            let mut user_query = msg.content.trim().to_string();
            break;
        }
    }
    if !user_query {
        return Err(anyhow::anyhow!("HTTPException(400, detail='No user message found')"));
    }
    let mut request_id = format!("ragchat-{}", /* uuid */ "00000000-0000-0000-0000-000000000000".to_string().hex[..12]);
    let mut created = std::time::SystemTime::now().duration_since(std::time::UNIX_EPOCH).unwrap().as_secs_f64().to_string().parse::<i64>().unwrap_or(0);
    let mut rag = _get_rag_integration();
    let mut context_text = "".to_string();
    let mut rag_sources = vec![];
    if (rag && rag.initialized) {
        // try:
        {
            let (mut context_text, mut raw_results) = rag.query_context(/* query= */ user_query, /* top_k= */ req.top_k, /* score_threshold= */ req.score_threshold).await;
            let mut rag_sources = raw_results.iter().map(|r| HashMap::from([("text".to_string(), r.get(&"text".to_string()).cloned().unwrap_or("".to_string())[..300]), ("score".to_string(), ((r.get(&"score".to_string()).cloned().unwrap_or(0) as f64) * 10f64.powi(3)).round() / 10f64.powi(3)), ("source".to_string(), r.get(&"source".to_string()).cloned().unwrap_or("unknown".to_string()))])).collect::<Vec<_>>();
        }
        // except Exception as e:
    }
    let mut focused_query = user_query;
    // try:
    {
        // TODO: from Core.prompt_focus import FocusMode, apply_focus
        if req.focus_mode {
            let mut _mode = FocusMode.from_string(req.focus_mode);
            if _mode != FocusMode.GENERAL {
                let (mut _focus_sys, mut focused_query) = apply_focus(_mode, user_query, req.system_prompt);
                if !req.system_prompt {
                    req.system_prompt = _focus_sys;
                }
            }
        }
    }
    // except ImportError as exc:
    let mut system_prompt = (req.system_prompt || "You are Zena, a helpful virtual assistant. Answer the user's question using ONLY the provided context. If the context doesn't contain the answer, say so honestly. Respond in the same language the user uses.".to_string());
    if context_text {
        system_prompt += format!("\n\n--- KNOWLEDGE BASE CONTEXT ---\n{}\n--- END CONTEXT ---", context_text);
    }
    let mut messages_for_llm = vec![HashMap::from([("role".to_string(), "system".to_string()), ("content".to_string(), system_prompt)])];
    for msg in req.messages.iter() {
        if (msg.role == "user".to_string() && msg.content.trim().to_string() == user_query) {
            messages_for_llm.push(HashMap::from([("role".to_string(), msg.role), ("content".to_string(), focused_query)]));
        } else {
            messages_for_llm.push(HashMap::from([("role".to_string(), msg.role), ("content".to_string(), msg.content)]));
        }
    }
    let mut state = _get_llm_state();
    let mut response_text = "".to_string();
    if (state && state::ready && state::adapter) {
        // try:
        {
            // TODO: from server::helpers import build_inference_request, estimate_tokens
            // TODO: from server::schemas import ChatCompletionRequest, ChatMessage
            let mut chat_req = ChatCompletionRequest(/* messages= */ messages_for_llm.iter().map(|m| ChatMessage(/* role= */ m["role".to_string()], /* content= */ m["content".to_string()])).collect::<Vec<_>>(), /* temperature= */ req.temperature, /* max_tokens= */ (req.max_tokens || 512), /* stream= */ false);
            let mut llm_req = build_inference_request(messages_for_llm, chat_req);
            let _ctx = state::inference_semaphore;
            {
                let mut result = asyncio.to_thread(state::adapter.generate, llm_req).await;
            }
            let mut response_text = if /* /* isinstance(result, str) */ */ true { result } else { result.to_string() };
            // try:
            {
                state::cache::put(messages_for_llm, req.temperature, (req.max_tokens || 512), response_text);
            }
            // except Exception as exc:
            let mut prompt_tokens = estimate_tokens(messages_for_llm);
            let mut completion_tokens = estimate_tokens(response_text);
            state::record_request(completion_tokens);
            JSONResponse(/* content= */ HashMap::from([("id".to_string(), request_id), ("object".to_string(), "rag.chat::completion".to_string()), ("created".to_string(), created), ("model".to_string(), /* getattr */ "rag-pipeline".to_string()), ("choices".to_string(), vec![HashMap::from([("index".to_string(), 0), ("message".to_string(), HashMap::from([("role".to_string(), "assistant".to_string()), ("content".to_string(), response_text)])), ("finish_reason".to_string(), "stop".to_string())])]), ("usage".to_string(), HashMap::from([("prompt_tokens".to_string(), prompt_tokens), ("completion_tokens".to_string(), completion_tokens), ("total_tokens".to_string(), (prompt_tokens + completion_tokens))])), ("rag_sources".to_string(), rag_sources)]))
        }
        // except Exception as e:
    }
    if context_text {
        let mut response_text = ("Based on the knowledge base, here is what I found:\n\n".to_string() + context_text[..2000]);
    } else {
        let mut response_text = "I don't have enough information to answer that question yet. Please make sure data has been loaded into the RAG pipeline.".to_string();
    }
    Ok(JSONResponse(/* content= */ HashMap::from([("id".to_string(), request_id), ("object".to_string(), "rag.chat::completion".to_string()), ("created".to_string(), created), ("model".to_string(), "rag-context-only".to_string()), ("choices".to_string(), vec![HashMap::from([("index".to_string(), 0), ("message".to_string(), HashMap::from([("role".to_string(), "assistant".to_string()), ("content".to_string(), response_text)])), ("finish_reason".to_string(), "stop".to_string())])]), ("usage".to_string(), HashMap::new()), ("rag_sources".to_string(), rag_sources)])))
}
